### PR TITLE
Adding checkpoint and config version

### PR DESCRIPTION
**Pull request by SamuelLarkin**
### PR Goal? 

Add a version number in a saved model's checkpoint and a version to EveryVoice's config files.

### Fixes? 

https://github.com/EveryVoiceTTS/EveryVoice/issues/559

### Feedback sought? 

Merge approval

### Priority? 

high

### Tests added? 

yes

### How to test? 

```sh
python -m unittest \
  everyvoice.tests.test_model.TestLoadingConfig \
  everyvoice.tests.test_model.TestLoadingModel \
  everyvoice.tests.test_cli
```

### Confidence? 

good

### Version change? 

yes for the schemas

### Related PRs? 

https://github.com/EveryVoiceTTS/FastSpeech2_lightning/pull/92
https://github.com/EveryVoiceTTS/HiFiGAN_iSTFT_lightning/pull/38
https://github.com/EveryVoiceTTS/DeepForcedAligner/pull/28

### CODE REVIEW

**Comment by joanise** [comment]
Darn, this is going to require a schema update, which normally means a minor version bump.

@roedoejet are the schemas published yet? If so, we'll need a minor bump to 0.2.0a0 instead of continuing with 0.1.0a4. Or maybe 0.2.0a4 to continue with our milestone naming? That's definitely not typical progression, but hey! sometimes we're stuck with stuff like that.

**Comment by joanise** [comment]
Version bump or not, I still think this is the right thing to do.

File: everyvoice/tests/test_model.py [hunk 1]
```diff
@@ -251,3 +253,185 @@ def test_model_is_not_a_vocoder(self):
...
+            self.assertIn("model_info", m.keys())
+            m["model_info"]["name"] = "BAD_TYPE"
+            torch.save(m, ckpt_fn)
+            m = torch.load(ckpt_fn)
+            self.assertIn("model_info", m.keys())
+            self.assertEqual(m["model_info"]["name"], "BAD_TYPE")
+            # self.assertEqual(m["model_info"]["version"], "1.0")
+            with self.assertRaisesRegex(
+                TypeError,
+                r"Wrong model type \(BAD_TYPE\), we are expecting a FastSpeech2",
```
Comment (roedoejet) [inline everyvoice/tests/test_model.py:302]:
```suggestion
                r"Wrong model type \(BAD_TYPE\), we are expecting a 'FastSpeech2' model",
```

File: everyvoice/tests/test_model.py [hunk 2]
```diff
@@ -251,3 +253,185 @@ def test_model_is_not_a_vocoder(self):
...
+            trainer.strategy.connect(model)
+            ckpt_fn = tmpdir_str + "/checkpoint.ckpt"
+            trainer.save_checkpoint(ckpt_fn)
+            m = torch.load(ckpt_fn)
+            self.assertIn("model_info", m.keys())
+            self.assertEqual(m["model_info"]["name"], FastSpeech2.__name__)
+            self.assertEqual(m["model_info"]["version"], BAD_VERSION)
+            with self.assertRaisesRegex(
+                ValueError,
+                r"Your model was created with a newer version of EveryVoice, please update your software.",
```
Comment (roedoejet) [inline everyvoice/tests/test_model.py:349]: I'm surprised that "BAD_VERSION" causes an error message that says "your model was created with a newer version of EveryVoice" - isn't "BAD_VERSION" just an unknown version? I would rather see messaging that indicates whether the provided version is newer/older or unknown.

**Review by roedoejet — COMMENTED** [review]
looks good to me! just a couple small things about the error messages

**Comment by roedoejet** [comment]
> Darn, this is going to require a schema update, which normally means a minor version bump.
> 
> @roedoejet are the schemas published yet? If so, we'll need a minor bump to 0.2.0a0 instead of continuing with 0.1.0a4. Or maybe 0.2.0a4 to continue with our milestone naming? That's definitely not typical progression, but hey! sometimes we're stuck with stuff like that.

right - yea if we release this as 0.1.0a4 then the schema will be broken, but I don't think it's that big of a deal actually. In the case where people don't change their configs to accommodate the new version key, the schema checker won't care, and if they create a new config with the version number, it'll just underline that value (if they have the yaml extension installed) but otherwise it shouldn't cause any problems. And, if they see the underline and remove the version number, it won't actually matter since we handle it right now with this PR anyways. Mind you, maybe we should get in the practise of bumping the schemas like this, in which case I think we should go to `0.2.0a0` and I'll change the milestone names.

**Comment by joanise** [comment]
> but I can't get the Vocoder loading to seem to call its own `check_and_upgrade_checkpoint` function.

I tried with a Vocoder I trained myself, and with `/gpfs/fs3c/nrc/dt/sgile/models/hifigan/hifigan_universal_v1_everyvoice.ckpt`, and in both cases `HiFiGANGenerator.check_and_upgrade_checkpoint()` does not appear to get called.

File: everyvoice/tests/test_cli.py [hunk 3]
```diff
@@ -177,38 +177,42 @@ def test_update_schema(self):
...
-                # Some objects will require a contact key
-                try:
-                    obj_instance = obj()
-                except ValidationError:
-                    obj_instance = obj(contact=dummy_contact)
-                self.assertIsNone(
-                    jsonschema.validate(
-                        json.loads(obj_instance.model_dump_json()),
-                        schema=schema,
+                with self.subTest(filename=filename, type=obj):
```
[the remark this answers is not in the document]
Comment (SamuelLarkin) [reply everyvoice/tests/test_cli.py:180]: We learn every day :D

**Review by joanise — APPROVED** [review]
Looks good, thanks for your patience with all the change requests!

**Pull request merged**